### PR TITLE
install header files

### PR DESCRIPTION
**Pull request by videlec**
Add a directive to install header files together with the library.

With these few lines added, the headers are installed in a directory `gmpy2/` inside the python installation.

Related to #121 

However, there would still remain the following problem: how can we find out where to find the gmpy2 repo that contains the header files?

### CODE REVIEW

**Comment by casevh** [comment]
Thanks for the patch.

I'll experiment with how to find header files this weekend.

**Comment by videlec** [comment]
Thanks.

Note that for modules that correspond to repository (and not a single .so file as gmpy2) there is an attribute `__path__` that points towards the directory, e.g.

    >>> import numpy
    >>> numpy.__path__
    ['/usr/lib/python3.6/site-packages/numpy']

(perhaps https://github.com/cython/cython/pull/1654 might be of some help)

File: setup.py [hunk 1]
```diff
@@ -251,6 +251,9 @@ def finalize_options(self):
       maintainer_email = "casevh@gmail.com",
       url = "http://code.google.com/p/gmpy/",
       description = "GMP/MPIR, MPFR, and MPC interface to Python 2.6+ and 3.x",
+      packages = ['gmpy2'],
+      package_dir = {'gmpy2': 'src'},
+      package_data = {'gmpy2': ['*.h']},
```
Comment (isuruf) [inline setup.py]: Why not use `headers = ['*.h']`?

**Comment by casevh** [comment]
I've pushed a couple of changes. Use of the Capsule API should only require gmpy2.h and not all the other header files. And setup.py is now using headers to specify gmpy2.h.

I left the package option in the code but commented out since they may (will?) be needed to distribute a gmpy2.pxd file.

Thanks for the review and comment.

**Comment by videlec** [comment]
Using @isuruf version, the header file gmpy2.h is now in a completely different location. On my computer it is `$HOME/.local/include/python3.6m/gmpy2/gmpy2.h`. How should I find out where this header has been installed (I need to detect it in my `setup.py` scripts)?

**Comment by isuruf** [comment]
@videlec, just use the header as `gmpy2/gmpy2.h`. Python's extension building system would add the `-I` invocations

**Comment by videlec** [comment]
@isuruf at least it does not work with Cython

    gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -fPIC -Ippl -I/usr/include/python3.6m -c ppl/gmpy2_wrap.c -o build/temp.linux-x86_64-3.6/ppl/gmpy2_wrap.o
    ppl/gmpy2_wrap.c:450:25: erreur fatale : gmpy2/gmpy2.h : Aucun fichier ou dossier de ce type
     #include "gmpy2/gmpy2.h"
                             ^
    compilation terminée.
    error: command 'gcc' failed with exit status 1

**Comment by isuruf** [comment]
Can you try `python setup.py install --user`?

**Comment by videlec** [comment]
This is the exact command I did to get this error... (more precisely `pip install . --user --upgrade --no-index`)

**Comment by videlec** [comment]
One can use the following to add "local" base for headers

    >>> from distutils.sysconfig import get_python_inc
    >>> import site
    >>> get_python_inc(prefix=site.USER_BASE)
    '/home/vincent/.local/include/python3.6m'

Though I think it is a bug with pip.

**Comment by videlec** [comment]
Sadly, my workaround seems to be not compatible with virtualenv, see eg [this failing travis build](https://travis-ci.org/videlec/pplpy/jobs/233522691)

**Comment by videlec** [comment]
Note that the numpy module has a nice function to deal with these kind of headaches

    >>> import numpy
    >>> numpy.get_include()
    '/usr/lib/python3.6/site-packages/numpy/core/include'

I think the simplest would be to add such function to gmpy2. What do you think?

**Comment by isuruf** [comment]
Ah, that explains why people don't use the `headers` option.

**Comment by videlec** [comment]
@casevh I don't understand why shipping only `gmpy2.h` might work. In this header you have

    #include "gmpy2_macros.h"
    #include "gmpy2_mpz.h"
    #include "gmpy2_xmpz.h"
    #include "gmpy2_mpq.h"
    #include "gmpy2_mpfr.h"
    #include "gmpy2_mpc.h"
    #include "gmpy2_context.h"
    #include "gmpy2_random.h"
    #include "gmpy2_cache.h"
    #include "gmpy2_misc.h"
    #include "gmpy2_binary.h"
    #include "gmpy2_convert.h"
    #include "gmpy2_convert_utils.h"
    #include "gmpy2_convert_gmp.h"
    #include "gmpy2_convert_mpfr.h"
    #include "gmpy2_convert_mpc.h"
    #include "gmpy2_mpz_divmod.h"
    #include "gmpy2_mpz_divmod2exp.h"
    #include "gmpy2_mpz_pack.h"
    #include "gmpy2_mpz_bitops.h"
    #include "gmpy2_mpz_inplace.h"
    #include "gmpy2_mpz_misc.h"
    #include "gmpy2_xmpz_inplace.h"
    #include "gmpy2_xmpz_misc.h"
    #include "gmpy2_mpq_misc.h"
    #include "gmpy2_mpfr_misc.h"
    #include "gmpy2_mpc_misc.h"
    #include "gmpy_mpz_lucas.h"
    #include "gmpy_mpz_prp.h"
    #include "gmpy2_abs.h"
    #include "gmpy2_add.h"
    #include "gmpy2_divmod.h"
    #include "gmpy2_floordiv.h"
    #include "gmpy2_minus.h"
    #include "gmpy2_mod.h"
    #include "gmpy2_mul.h"
    #include "gmpy2_plus.h"
    #include "gmpy2_pow.h"
    #include "gmpy2_sub.h"
    #include "gmpy2_truediv.h"
    #include "gmpy2_math.h"
    #include "gmpy2_const.h"
    #include "gmpy2_square.h"
    #include "gmpy2_format.h"
    #include "gmpy2_hash.h"
    #include "gmpy2_fused.h"
    #include "gmpy2_muldiv_2exp.h"
    #include "gmpy2_predicate.h"
    #include "gmpy2_sign.h"
    #include "gmpy2_richcompare.h"
    #  include "gmpy2_vector.h"

where these files are supposed to be found?

**Comment by casevh** [comment]
The sheer number of source/include files was a side-effect of the refactoring that was done to add thread-safe contexts. The majority of the include files are not related to the public C API. I moved the portions that were needed for the public API to gmpy2.h directly. Note that the other header files are only imported when GMPY2_MODULE is defined before importing gmpy2.h

**Pull request merged**